storage/shard.go

```diff
@@ -297,8 +309,54 @@ func (s *dbShard) IsBlockRetrievable(blockStart time.Time) bool {
 		flushState.Status))
 }
 
+func (s *dbShard) OnRetrieveBlock(
+	id ts.ID,
+	startTime time.Time,
+	segment ts.Segment,
+) {
+	s.RLock()
+	entry, _, err := s.lookupEntryWithLock(id)
+	if entry != nil {
+		entry.incrementReaderWriterCount()
+	}
+	s.RUnlock()
+	defer func() {
```

Comment:
Not sure how hot this will be in the grand schema of things, but could avoid the defer with a helper function like

```golang
onRetrieveBlockCleanup(entry *dbShardEntry, id ts.ID) {
  if entry != nil {
    entry.decrementReaderWriterCount()
  }
  id.Finalize()
}
```

Reply:
Could, however since go 1.8 it's pretty cheap, roughly 25-30 nanoseconds, which is fine since this only happens on reads which is already ripe for optimization in other places first.